### PR TITLE
feat: Telemetry reporting

### PR DESCRIPTION
**Pull request by michalbali256**

### CODE REVIEW

File: clients/vscode-hlasmplugin/src/extension.ts [hunk 1]
```diff
@@ -76,10 +105,13 @@ export async function activate(context: vscode.ExtensionContext) {
 
     if (serverVariant === 'native')
         startCheckingNativeClient(hlasmpluginClient);
-
+
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/extension.ts:108]:
```suggestion

```

File: language_server/src/dap/dap_server.cpp [hunk 2]
```diff
@@ -71,6 +71,7 @@ void server::message_received(const json& message)
         if (message.at("type") != "request")
         {
             LOG_WARNING(std::string("Invalid message receive: ") + message.dump());
+            telemetry_error("lsp_server/invalid_message");
```
Comment (slavek-kucera) [inline language_server/src/dap/dap_server.cpp:74]:
```suggestion
            telemetry_error("dap_server/invalid_message");
```

File: language_server/src/server.h [hunk 3]
```diff
@@ -64,11 +67,25 @@ class server : public response_provider
...
     // Calls a method that is registered in methods_ with the specified name with arguments and id of request.
     void call_method(const std::string& method, const json& id, const json& args);
+
+    struct telemetry_metrics_info
+    {
+        json properties;
+        json metrics;
+    };
+
+    telemetry_metrics_info virtual get_telemetry_details();
```
Comment (slavek-kucera) [inline language_server/src/server.h:83]: This seems a bit strange... how is the method supposed to know to which request the details are being provided?

File: language_server/src/lsp/lsp_server.h [hunk 4]
```diff
@@ -41,6 +43,11 @@ class server final : public hlasm_plugin::language_server::server,
     // Parses LSP (JSON RPC) message and calls corresponding method.
     void message_received(const json& message) override;
 
+    // Inherited via json_sink
+    // Takes json with telemetry info and sends it through LSP
+    void write(const nlohmann::json&) override;
+    void write(nlohmann::json&&) override;
+
```
Comment (slavek-kucera) [inline language_server/src/lsp/lsp_server.h:50]: wouldn't it be better to have a dedicated interface that takes a telemetry_info type?

**Review by slavek-kucera — CHANGES_REQUESTED** [review]
Some of the code smells should probably be addressed.

File: clients/vscode-hlasmplugin/src/extension.ts [hunk 5]
```diff
@@ -23,22 +23,46 @@ import { CustomEditorCommands } from './customEditorCommands';
...
  */
 export async function activate(context: vscode.ExtensionContext) {
+    const serverVariant = getConfig<ServerVariant>('serverVariant', 'native');
+    
+    var telemetry = new Telemetry();
+    context.subscriptions.push(telemetry);
+    
+    // setTimeout is needed, because telemetry initialization is asynchronous
+    // and AFAIK no event in the API is exposed to send the activation telemetry event
+    setTimeout(() => {telemetry.reportEvent("hlasm.activated", {server_variant:serverVariant.toString()});}, 1000);
```
[the remark this answers is not in the document]
Comment (michalbali256) [reply clients/vscode-hlasmplugin/src/extension.ts:59]: I don't think so. I wanted to have an event that would trigger right after activation and would not depend on e.g. ability to run native executables.

BTW creators of vscode-extension-telemetry acknowledged that this is a shortfall of the plugin and plan to either introduce an event or buffer where such event would be stored before they can be sent. https://github.com/microsoft/vscode-extension-telemetry/issues/76

File: clients/vscode-hlasmplugin/src/extension.ts [hunk 6]
```diff
@@ -50,16 +74,21 @@ export async function activate(context: vscode.ExtensionContext) {
             code2Protocol: (value: vscode.Uri) => value.toString(),
             protocol2Code: (value: string) =>
                 vscode.Uri.file((vscode.Uri.parse(value).fsPath))
-        }
+        },
+        errorHandler: clientErrorHandler
```
[the remark this answers is not in the document]
Comment (michalbali256) [reply clients/vscode-hlasmplugin/src/extension.ts:78]: Yes it does

File: language_server/src/server.h (hunk 3 above)
Comment (michalbali256) [reply language_server/src/server.h]: I agree it is a bit strange... the details are provided for the last request.

It fetches the details for requests with `telemetry_log_level::LOG_WITH_METRICS`, which is actually only one request (didOpen).

I only wanted the `lsp_server` to have the metadata consumer, but at the same time I wanted to log the requests in the `server::call_method`, method, which is common for lsp and dap server. Thats why I ended up with this virtual method.

**Review by slavek-kucera — APPROVED** [review]

**Pull request merged**